File: frontend/src/api/portfolio.ts
import api from './index'

export interface AssetOverview {
  total_value: number
  total_balance: number
  total_market_value: number
  total_unrealized_pnl: number
  total_realized_pnl: number
  total_pnl: number
  total_pnl_pct: number
  account_count: number
}

export interface MarketAllocation {
  market: string
  market_value: number
  balance: number
  total_value: number
  percentage: number
  account_count: number
}

export interface HoldingItem {
  symbol: string
  name: string | null
  market: string
  account_name: string
  account_id: string
  quantity: number
  avg_cost: number
  current_price: number
  market_value: number
  unrealized_pnl: number
  unrealized_pnl_pct: number
  realized_pnl: number
}

export interface CashFlowItem {
  id: string
  account_name: string
  symbol: string
  side: string
  quantity: number
  price: number
  commission: number
  amount: number
  filled_at: string
}

export async function getOverview(): Promise<AssetOverview> {
  const resp = await api.get<AssetOverview>('/v1/portfolio/overview')
  return resp.data
}

export async function getAllocation(): Promise<MarketAllocation[]> {
  const resp = await api.get<MarketAllocation[]>('/v1/portfolio/allocation')
  return resp.data
}

export async function getHoldings(): Promise<HoldingItem[]> {
  const resp = await api.get<HoldingItem[]>('/v1/portfolio/holdings')
  return resp.data
}

export interface PerformancePoint {
  date: string
  total_value: number
}

export async function getPerformance(days?: number): Promise<PerformancePoint[]> {
  const resp = await api.get<PerformancePoint[]>('/v1/portfolio/performance', {
    params: { days },
  })
  return resp.data
}

export async function getCashFlows(limit?: number): Promise<CashFlowItem[]> {
  const resp = await api.get<CashFlowItem[]>('/v1/portfolio/cash-flows', {
    params: { limit },
  })
  return resp.data
}

File: backend/app/schemas/portfolio.py
"""Portfolio / asset dashboard schemas."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class AssetOverview(BaseModel):
    """Aggregated asset summary across all accounts."""

    total_value: Decimal
    total_balance: Decimal
    total_market_value: Decimal
    total_unrealized_pnl: Decimal
    total_realized_pnl: Decimal
    total_pnl: Decimal
    total_pnl_pct: Decimal
    account_count: int


class MarketAllocation(BaseModel):
    """Allocation breakdown by market."""

    market: str
    market_value: Decimal
    balance: Decimal
    total_value: Decimal
    percentage: Decimal
    account_count: int


class HoldingItem(BaseModel):
    """Single holding across accounts."""

    symbol: str
    name: str | None
    market: str
    account_name: str
    account_id: uuid.UUID
    quantity: int
    avg_cost: Decimal
    current_price: Decimal
    market_value: Decimal
    unrealized_pnl: Decimal
    unrealized_pnl_pct: Decimal
    realized_pnl: Decimal


class PerformancePoint(BaseModel):
    """Single data point on the performance curve."""

    date: date
    total_value: Decimal


class CashFlowItem(BaseModel):
    """A single cash flow event (order fill)."""

    id: uuid.UUID
    account_name: str
    symbol: str
    side: str
    quantity: int
    price: Decimal
    commission: Decimal
    amount: Decimal
    filled_at: datetime

    model_config = {"from_attributes": True}
